File: perl/rp_list_contacts.pl
#!/usr/bin/perl

# Program: rp_list_contacts.pl
# Usage: ./rp_list_contacts.pl
# Purpose: perl language sample client program for RackPing Monitoring API 2.0
# Version: 1.0
# Copyright: RackPing USA 2020
# Env: Perl5
# Returns: exit status is non-zero on failure
# Note: First set the envariables with: source ../set.sh

use strict;
use diagnostics;

use MIME::Base64 qw(encode_base64);
use LWP::UserAgent;

   use constant TIMEOUT_SEC => 5.0;

   my $url = $ENV{'RP_SCHEME'} . $ENV{'RP_DOMAIN'} . $ENV{'RP_BASE_URL'};

   my $api_key  = $ENV{'RP_API_KEY'}  // '';
   my $user     = $ENV{'RP_USER'}     // '';
   my $password = $ENV{'RP_PASSWORD'} // '';
   my $timeout  = $ENV{'RP_TIMEOUT'}  || TIMEOUT_SEC;
   my $useragent = $ENV{'RP_USERAGENT'} // '';
   my $DEBUG    = $ENV{'RP_DEBUG'}    || 0;

   if ($api_key eq '') {
      print "error: do source ../set.sh to set the envariables.\n";
      exit 1;
   }

### start of contact settings

### end of contact settings

   my @headers = (
      'Accept'         => 'application/json',
      'Accept-Charset' => 'utf-8',
      'App-key'        => "$api_key",
      'Authorization'  => 'Basic ' . encode_base64("$user:$password", ''),
      'Content-type'   => 'application/json',
      'User-Agent'     => $useragent,
   );

# Create a contact agent object
   my $ua = LWP::UserAgent->new;
   $ua->from('rackping@example.com');
   $ua->timeout($timeout); # 411 - not reliable for https requests?

# Create a request for a list of contacts
   my $req = HTTP::Request->new(GET => $url . '/contacts');
   $req->header(@headers);

# Pass request to the contact agent and get a response back
   my $res = $ua->request($req);

   print STDERR "info: show list of contacts\n" if $DEBUG;

# Check the outcome of the response
   if ($res->is_success) {
      print $res->content, "\n";
   }
   else {
      print $res->status_line, "\n";
      exit 1;
   }

   exit 0;

File: rust/rp_list_contacts.rs
// Program: rp_list_contacts.rs
// Usage: cargo run
// Date: 2021 02 06
// Purpose: rust language sample client program for RackPing Monitoring API 2.0
// Version: 1.0
// Copyright: RackPing USA 2021
// Env: rust (customize source code for your version of rust)
// Returns: exit status is non-zero on failure
// Notes:
// - first do: source ../set.sh
// - this alpha (non-functional) sample code is a template for advanced rust programmers to customize for their version of rust:
//   - install rust and find out the version
//   - decide if you want async or blocking call to get
//   - update the main() definition with the correct signature if async
//   - update the call to get
//   - update the response println!()
//   - See:
//     - https://github.com/seanmonstar/reqwest
//     - https://github.com/seanmonstar/reqwest/blob/master/examples/blocking.rs
//     - https://docs.rs/reqwest/0.9.20/reqwest/struct.Response.html

use reqwest;
use std::env;
use std::process;
use std::io::Read;
// use serde_json;
// use std::error::Error;
// use std::result::Result;
// use std::time::Duration;
// use std::collections::HashMap;

fn main() {
   let username   = env::var("RP_USER").unwrap();
   let password   = env::var("RP_PASSWORD").unwrap();
   let api_key    = env::var("RP_API_KEY").unwrap();
   let user_agent = env::var("RP_USERAGENT").unwrap();
  
   if username == "" {
      println!("error: do 'source ../set.sh' first.\n");
      process::exit(1);
   }

   let domain     = env::var("RP_DOMAIN").unwrap();
   let base_url   = env::var("RP_BASE_URL").unwrap();
   let scheme     = env::var("RP_SCHEME").unwrap();

   let _timeout   = env::var("RP_TIMEOUT").unwrap().parse::<u64>().unwrap();
   let _redirects = env::var("RP_REDIRECTS").unwrap().parse::<u32>().unwrap();
   let debug      = env::var("RP_DEBUG").unwrap().parse::<u8>().unwrap();

   let url        = scheme.to_string() + &domain.to_string() + &base_url.to_string() + "/contacts";

   println!("Get list of contacts\n");

   let client  = reqwest::blocking::Client::new();

   let mut resp = client.get(&url)
       .header("User-agent", user_agent)
       .header("app-key", api_key)
       .basic_auth(username, Some(String::from(password)))
//     .json::<HashMap<String, String>>();
//     .json::<serde_json::Value>();
//     .timeout(Duration::from_secs(_timeout))
       .send()
       .unwrap();

    let mut body = String::new();
    resp.read_to_string(&mut body).unwrap();

    if debug == 1 {
       println!("Status: {}", resp.status());
       println!("Headers:\n{:#?}", resp.headers());
    }

    println!("{}", body);
}

